realm_domains: Move realm domains table to organization permissions.

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
The Realm domains table used to be in a modal. This PR moves that table to be in organization permissions of organization settings

**Testing Plan:** <!-- How have you tested? -->
Ran automated tests.
Manually tested UI to cover everything possible in that area.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![realm-add](https://user-images.githubusercontent.com/44665669/83185088-10cfea00-a148-11ea-9534-d11bd9219926.png)
![2](https://user-images.githubusercontent.com/44665669/83185386-82a83380-a148-11ea-9e0f-65c6844c7794.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

chdinesh1089 committed:
realm_domains: Restrict change of email restrictions automatically.

We want the admin to always be able to express their preference
on whether emails are restricted to domains, even when there aren't
actually any domains. The admin may want to set the flag now and
add domains later. In fact, our UI is sorta targeted at this workflow.

Before this change, we were forcing the flag to `False` when a user
removed the last domain. But we weren't even forcing the flag to
`False` in all situations (for eg. when admin restricts emails but
doesn't configure any domain). So the old code change made in
8ae54ddb99 didn't simplify any logic elsewhere in the system.

Now we remove that old logic and just base `emails_restricted_to_domain`
off of the user's last choice.

Also resolves the auto hiding of realm domains table that'll be shifted
to be on organization permissions page in the next commit.

Thanks to Steve Howell for the clear and concise commit message.

chdinesh1089 committed:
realm_domains: Move realm domains table to organization permissions.

The Realm domains table used to be in a modal. This commit moves
that table to be in organization permissions of organization settings.

This change is prerequisite to have per-domain setting to require
invites when new members join. When the `invite_required` field is
added to `RealmDomain` we'd need to make it visually clear to the
user about the state of table when global changes are made to avoid
any confusion.

chdinesh1089 commented on static/js/settings_org.js:
```diff
@@ -276,12 +276,16 @@ exports.populate_realm_domains = function (realm_domains) {
-    $("#allowed_domains_label").text(i18n.t("Allowed domains: __domains__", {domains: domains}));
 
-    const realm_domains_table_body = $("#realm_domains_table tbody").expectOne();
+    const realm_domains_table_body = $("#realm_domains_table tbody");
```
Made this change because when a normal user logins `#realm_domains_table` is not expected to be present.

aero31aero replied:
Good catch! We should have a better fix than this, as this removes a sanity check from the code. Maybe `if(!is_admin) return;` somewhere above this line.

chdinesh1089 replied:
Added that `if` in the top like this because we do not  want to populate realm domains table for normal users
```
exports.populate_realm_domains = function (realm_domains) {
    if (!meta.loaded || !page_params.is_admin) {
        return;
    }

timabbott commented:
@chdinesh1089 thanks for working on this!  I think we need to do a bit more work to make this UI acceptable:
* The use of the old-style green banner feedback and general confirmation logic for the modal is inconsistent with the rest of this page.  
* The "Waiting period" setting ends up out of place in a confusing way.
* The heading would be better as an h4 rather than an h3, unless we're hoping to make it a setting heading.

To resolve these:
* Let's add a prep commit that moves "Waiting period before new members turn into full members " to the "User identity" section; while that's a bit off from what we want long-term, I think given the restructuring we're doing to everything in "Joining the organization", we'll want it in another place for now.  Be sure to update the Help Center in that commit.
* Let's add a next series of commits that work on making the save behavior for changes to the "Allowed domains" form consistent with how the other settings in "Joining the organization" work.

chdinesh1089 committed:
org_permissions: Move waiting period setting to the top.

When realm domains modal is moved to be on organization permissions
page, "Waiting period before new members turn into full members"
setting ends up in out of place in a confusing way. So, this
commit moves that setting to the top in "Joining the organization"
block to make it not look odd when "Admin domains" table is
moved to be on organization permissions page.

chdinesh1089 committed:
realm_domains: Change save behaviour of realm domains table.

Removes the inconsistent saved/failed status it had when the
this was displayed in a modal. Now, since we moved it to
organization permissions page, we want it to be consistent
with save behaviour of the other sections in the page.

We do not show "Save", "Discard" option for these and
instead save directly and show the staus.

chdinesh1089 committed:
org_permissions: Split out domain_restrictions section.

In this commit, we split out a section "Domain restrictions"
from the org_join section. This new section contains settings
for restricting who can join the realm.

This is done to remove the need for a separate modal to
restrict joining to specific domains.

Thanks to Rohitt Vashishtha for the clear commit message.

chdinesh1089 commented:
Made the changes and modified tests as needed. This is how it looks now.
![domains_table](https://user-images.githubusercontent.com/44665669/84658036-dc737080-af32-11ea-92bc-a9b7a7565d4e.gif)


(link to CircleCI: https://circleci.com/gh/chdinesh1089/zulip/tree/domains-table)

aero31aero commented on static/js/settings_org.js:
```diff
@@ -244,8 +244,12 @@ function set_message_retention_setting_dropdown() {
     $("#id_realm_org_join_restrictions").val(value);
-    change_element_block_display_property('allowed_domains_label',
-                                          value === 'only_selected_domain');
+    const zero_domains_warning = $('#zero_domains_warning');
```
Can we not make this new warning reuse that show/hide scheme that `change_element_block_display_property` provides?

chdinesh1089 replied:
That function hides the block in which `#zero_domains_warning` is present. so, it's hiding the whole table.

aero31aero commented on static/js/settings_org.js:
```diff
@@ -844,24 +850,14 @@ exports.build_page = function () {
         const org_join_restrictions = e.target.value;
-        const node = $("#allowed_domains_label").parent();
+        const node = $("#zero_domains_warning").parent();
         if (org_join_restrictions === 'only_selected_domain') {
```
This is the third time I'm seeing this check for showing/hiding the warning. Maybe this can be extracted to a small `show_hide_domain_warning(org_join_restrictions);`.

chdinesh1089 replied:
I removed this whole block in the next commit introducing "Domain restrictions" section, it's still good to have a function for this, will add a function then.

aero31aero replied:
Sorry, I was reviewing this commit by commit. For 2 cases, it doesn't seem that important.

chdinesh1089 replied:
Okay, didn't change it.

aero31aero commented on static/templates/settings/realm_domains_table.hbs:
```diff
@@ -0,0 +1,24 @@
+    </thead>
+    <tbody>
+    </tbody>
+    {{#if is_admin }}
```
Shouldn't we just hide the whole table for non-admins, instead of showing a blank table?

chdinesh1089 replied:
We do not show the whole of organization permissions for normal users. I think it's redundant here.(I renamed file from realm_domains_modal.hbs and remove little code, it was from that page. Thought it's probably needed)
Can I remove that?

aero31aero replied:
Lets remove this then. We should strive to keep templates free of logic where possible.

chdinesh1089 replied:
Removed.

aero31aero commented on static/templates/settings/organization_permissions_admin.hbs:
```diff
@@ -36,20 +36,29 @@
                         <option value="no_restriction">{{t "No restrictions" }}</option>
                         <option value="no_disposable_email">{{t "Don’t allow disposable email addresses" }}</option>
-                        <option value="only_selected_domain">{{t "Restrict to a list of domains" }}</option>
+                        <option value="only_selected_domain">{{t "Restrict to the below list of domains" }}</option>
```
I think `Restrict to the domains listed below` flows better.

chdinesh1089 replied:
changed it.

aero31aero commented on static/templates/settings/organization_permissions_admin.hbs:
```diff
@@ -36,20 +36,29 @@
                     <p id="zero_domains_warning">There are no domains configured now. This'd restrict anyone from registering.</p>
                     {{> realm_domains_table }}
                 </div>
+
```
Intentional whiespace?

chdinesh1089 replied:
Ohh, no that was unintentional. I'll remove it.

chdinesh1089 replied:
Removed it.

chdinesh1089 committed:
realm_domains: Remove redundant `if` in realm_domains_table.hbs.

We hide the whole of organization permissions section for normal users.
So, this check isn't necessary.

chdinesh1089 commented:
Thanks for reviewing this @aero31aero. Made the suggested changes.

chdinesh1089 commented on templates/zerver/help/invite-new-users.md:
```diff
@@ -34,8 +34,10 @@ invitation, but require them to authenticate via LDAP.
 
 1. Set **Are invitations required for joining the organization** to **No**.
 
+1. Find the section **Domain restrictions**.
+
```
Not sure if it is right to say this as this is just below the "Joining the organization" section and will be clearly visible but I think we should make it clear that "Restrict email domains of new users?" is in different section. Is it fine or do we need to rephrase this or consider not adding this line?

chdinesh1089 commented:
Rebased and fixed conflicts. Could you do a review @timabbott ?
